core/src/entry_verification_service.rs

```diff
@@ -0,0 +1,162 @@
+    ) -> Result<bool, WaitForResultError> {
+        self.commands
+            .send(Command::WaitingFor { slot, data_set })
+            .map_err(|_| WaitForResultError::WorkerTerminated)?;
+
+        let wait_start = Instant::now();
+        match self.results.check_and_wait_until(deadline, |results| {
+            if slot < results.starting_at {
+                return Some(false);
+            }
+
+            match results.slots.get(&slot) {
+                None => None,
+                Some(SlotVerificationStatus::Failed) => Some(false),
+                Some(SlotVerificationStatus::Partial { verified_up_to })
+                    if data_set < *verified_up_to =>
+                {
+                    Some(true)
+                }
+                Some(SlotVerificationStatus::Partial { .. }) => None,
+                Some(SlotVerificationStatus::Valid) => Some(true),
+            }
+        }) {
+            Some(valid) => Ok(valid),
+            None => Err(WaitForResultError::Timeout {
+                slot,
+                data_set,
+                waited_for: Instant::now() - wait_start,
+            }),
+        }
+    }
+
+    pub fn cleanup(&self, earlier_than: Slot) -> Result<(), CleanupError> {
+        self.commands
+            .send(Command::Cleanup { earlier_than })
+            .map_err(|_| CleanupError::WorkerTerminated)
+    }
+}
+
+/// We index data sets using index of the first shred that contains them.  This type represents this
+/// kind of index - it identifies a data set in a slot.
+#[derive(PartialEq, Eq, PartialOrd, Ord, Hash, Debug, Display, Clone, Copy, From, Into)]
+pub struct DataSetFirstShred(u32);
+
+impl DataSetFirstShred {
+    pub fn is_first_in_slot(&self) -> bool {
+        self.0 == 0
+    }
+}
+
+#[derive(PartialEq, Eq, Debug, Clone, Copy)]
+enum SlotVerificationStatus {
+    /// Slot verification failed.
+    Failed,
+    /// Data sets up to, but not including the specified one have been verified.
+    Partial { verified_up_to: DataSetFirstShred },
+    /// All data sets in the slot have been verified.
+    Valid,
+}
+
+/// Results that are shared between the worker and the API of the service.
+/// Wrapping it in [`SharedWrapper`] allows the API to wait for the worker, without actively
+/// checking for the state to be updated.
+struct Results {
+    /// First slot that we are going to process and, possibly, have any results for.
+    ///
+    /// `EntryVerificationService::cleanup()` shifts this value forward.  And all the other
+    /// operations are only valid, when called for this or later slots.
+    starting_at: Slot,
+
+    slots: HashMap<Slot, SlotVerificationStatus>,
+}
+
+impl Default for Results {
+    fn default() -> Self {
+        Self {
+            starting_at: 0,
+            slots: HashMap::new(),
+        }
+    }
+}
+
+impl Results {
+    fn set_slot_status(&mut self, slot: Slot, status: SlotVerificationStatus) {
+        if slot < self.starting_at {
```

Comment:
Is this expected or an error condition?

Reply:
Addressed this in the thread above.
Let's continue there.